qiime2/core/type/signature.py

```diff
@@ -484,6 +485,25 @@ def coerce_given_outputs(self, output_views, output_types, scope,
                         name=name, key=key, idx=idx, size=size)
                     output[key] = self._create_output_artifact(
                         provenance, collection_name, scope, spec, view)
+            elif isinstance(get_origin(spec.view_type), type(Union)):
+                union_resolved = False
+                for arg in get_args(spec.view_type):
+                    if isinstance(output_view, arg):
+                        # we need a new spec with the correct view type
+                        spec = ParameterSpec(
+                            qiime_type=spec.qiime_type, view_type=arg,
```

Comment:
Having not looked very much into it, should this be something we handle in `transform.py`? It's a scary place, but it seems like where I would expect to deconstruct the union.

Reply:
Hey @ebolyen, resurfacing this thread - any thoughts on the above?